src/platforms/x11/X11_resources.h

```diff
@@ -38,19 +40,31 @@ int mir_x11_error_handler(Display* dpy, XErrorEvent* eev);
 class X11Resources
 {
 public:
+    class VirtualOutput
+    {
+    public:
+        VirtualOutput() = default;
+        virtual ~VirtualOutput() = default;
+        VirtualOutput(VirtualOutput const&) = delete;
+        VirtualOutput& operator=(VirtualOutput const&) = delete;
+
+        virtual auto configuration() const -> graphics::DisplayConfigurationOutput const& = 0;
+    };
+
     auto get_conn() -> std::shared_ptr<::Display>;
-    void set_output_config_for_win(
-        Window win,
-        std::weak_ptr<graphics::DisplayConfigurationOutput const> configuration);
-    void clear_output_config_for_win(Window win);
-    auto get_output_config_for_win(Window win)
-        -> std::experimental::optional<graphics::DisplayConfigurationOutput const* const>;
+
+    void set_set_output_for_window(Window win, VirtualOutput* output);
+    void clear_output_for_window(Window win);
+
+    /// X11Resources's mutex stays locked while the provided function runs
+    void with_output_for_window(Window win, std::function<void(std::optional<VirtualOutput const*> output)> fn);
 
     static X11Resources instance;
 
 private:
+    std::mutex mutex;
```

Comment:
This is for the future? It is not needed at present as all the access is during the single-threaded initialization phase of execution.